lib/config/loader.js

```diff
@@ -196,6 +209,62 @@ module.exports = function (userCfgPath) {
 					default: userConfig.login && userConfig.login.twitch && userConfig.login.twitch.enabled ? null : []
 				}
 			},
+			discord: {
+				enabled: {
+					doc: 'Whether to enable Discord authentication.',
+					format: Boolean,
+					default: false
+				},
+				clientID: {
+					doc: 'A Discord application ClientID https://discord.com/developers/applications/',
+					format: String,
+
+					// This will throw if the user does not provide a value, but only if Discord auth is enabled.
+					default: userConfig.login && userConfig.login.discord && userConfig.login.discord.enabled ? null : ''
+				},
+				clientSecret: {
+					doc: 'A Discord application ClientSecret https://discord.com/developers/applications/',
+					format: String,
+
+					// This will throw if the user does not provide a value, but only if Discord auth is enabled.
+					default: userConfig.login && userConfig.login.discord && userConfig.login.discord.enabled ? null : ''
+				},
+				scope: {
+					doc: 'A space-separated string of Discord application permissions.',
+					format: String,
+					default: 'identify'
+				},
+				allowedUserIDs: {
+					doc: 'Which Discord IDs to allow',
+					format: VALIDATE_STRING_ARRAY,
+
+					// This will throw if the user does not provide a value and is not using allowedGuilds, but only if Discord auth is enabled.
+					default: userConfig.login && userConfig.login.discord && userConfig.login.discord.enabled && !userConfig.login.discord.allowedGuilds ? null : []
+				},
+				allowedGuilds: {
+					doc: 'Which servers to allow users from',
+					format: 'discord-guild-array',
+					default: userConfig.login && userConfig.login.discord && userConfig.login.discord.enabled && !userConfig.login.discord.allowedUserIDs ? null : [],
+					children: {
+						guildID: {
+							doc: 'Users in this Discord Server are allowed to log in',
+							format: String,
+							default: userConfig.login && userConfig.login.discord && userConfig.login.discord.enabled && userConfig.login.discord.allowedGuilds ? null : ''
+						},
+						allowedRoleIDs: {
+							doc: 'Additionally require one of the roles on the server to log in',
+							format: VALIDATE_STRING_ARRAY,
+							default: []
+						},
+						guildBotToken: {
+							doc: 'Discord bot token, needed if allowedRoleIDs is used',
+							format: String,
+							// TODO: Make this do a proper check
```

Comment:
@Hoishin I'm not sure how I would write a check here that depends on if the same object has allowedRoleIDs, do you know how to do that?